format/midi/midi.go

```diff
@@ -0,0 +1,236 @@
+package midi
+
+// https://www.midi.org/specifications/item/the-midi-1-0-specification
+
+import (
+	"bytes"
+	"embed"
+	"encoding/binary"
+	"fmt"
+
+	"github.com/wader/fq/format"
+	"github.com/wader/fq/pkg/decode"
+	"github.com/wader/fq/pkg/interp"
+)
+
+type context struct {
+	tick    uint64
+	running uint8
+	casio   bool
+}
+
+//go:embed midi.md
+var midiFS embed.FS
+
+func init() {
+	interp.RegisterFormat(
+		format.MIDI,
+		&decode.Format{
+			Description: "Standard MIDI file",
+			Groups:      []*decode.Group{format.Probe},
+			DecodeFn:    decodeMIDI,
+		})
+
+	interp.RegisterFS(midiFS)
+}
+
+func decodeMIDI(d *decode.D) any {
+	d.Endian = decode.BigEndian
+
+	// ... decode header
+	d.FieldStruct("header", decodeMThd)
+
+	// ... decode tracks
+	d.FieldArray("tracks", func(d *decode.D) {
+		for d.BitsLeft() > 0 {
+			if err := skipTo(d, "MTrk"); err != nil {
```

Comment:
Not garbage no - it's a little bit ambiguous but at the very least it has to be a valid chunk with a 4 character tag and a length. The theory being that it would allow extensions and customisations without breaking existing MIDI file decoders.

**But** .. I've only ever seen this referenced once and can't remember having ever come across a MIDI file that had anything other than _MThd_ and _MTrk_ chunks, so I've removed it. If anybody ever raises an issue and can supply a MIDI file with _other_ chunks it's easy enough to put back.

Removed in [9f057b6](https://github.com/wader/fq/pull/1004/commits/9f057b6760bdec76fad892e81f3eb4a6cba34f37).

Reply:
I see, could it make sense to have a case that decodes unknown tags as just tag+data?

Reply:
Ok, just to make sure I dug out the [spec](https://midi.org/standard-midi-files) and this is what it says:

> MIDI Files are made up of chunks. Each chunk has a 4-character type and a 32-bit length,
> which is the number of bytes in the chunk. This structure allows future chunk types to be
> designed which may easily be ignored if encountered by a program written before the chunk
> type is introduced. Your programs should expect alien chunks and treat them as if they
> weren't there.

So _tag_ + _data_ makes sense - it's not a string so I've also included the length:
```
    |                                               |                |  tracks[0:5]: 0xe-0x56 (72)
    |                                               |                |    [0]{}: other 0xe-0x1e (16)
0x00|                                          4d 54|              MT|      tag: "MTrx" 0xe-0x12 (4)
0x10|72 78                                          |rx              |
0x10|      00 00 00 08                              |  ....          |      length: 8 0x12-0x16 (4)
0x10|                  01 23 45 67 89 ab cd ef      |      .#Eg....  |      data: raw bits 0x16-0x1e (8)
    |                                               |                |    [1]{}: track 0x1e-0x2a (12)
0x10|                                          4d 54|              MT|      tag: "MTrk" 0x1e-0x22 (4)
```

Not sure that it's really correct to call the section "_tracks_" any longer though - what do you think?

Reply:
👍 reclutant to call them chunks?

Reply:
Went with _content_ in the end - as in _header_ + _content_. It seemed to make the most sense :-)

Fixed in [9c7f7f9](https://github.com/wader/fq/pull/1004/commits/9c7f7f961a94561c8ef4b532d4ec7ab07c001726).